### PR TITLE
Improve torch.cuda.amp type hints

### LINKED ISSUE #108629: Improve torch.cuda.amp type hints
### 🐛 Describe the bug

1. `pyright` complains that none of the following classes or methods are exported by their modules, despite their docs pages being for that specific module:

* [`torch.autocast`](https://pytorch.org/docs/stable/amp.html#torch.autocast)
* [`torch.cuda.amp.GradScaler`](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler)
* [`torch.cuda.amp.autocast`](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.autocast)
* [`torch.cuda.amp.custom_fwd`](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.custom_fwd)
* [`torch.cuda.amp.custom_bwd`](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.custom_bwd)

```python
import torch
import torch.cuda.amp

torch.autocast  # "autocast" is not exported from module "torch"

torch.cuda.amp.GradScaler()  # "GradScaler" is not exported from module "torch.cuda.amp"

torch.cuda.amp.autocast  # "autocast" is not exported from module "torch.cuda.amp"

torch.cuda.amp.custom_fwd  # "custom_fwd" is not exported from module "torch.cuda.amp"

torch.cuda.amp.custom_bwd  # "custom_bwd" is not exported from module "torch.cuda.amp"

```

This can be fixed by adding them to their modules' `__all__`.

2. `torch.autocast.__call__` is not type hinted and gives unhelpful inferred types:
```python
import torch

def add(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
    return x + y

wrapped_add = torch.autocast(device_type="cuda")(add)

reveal_type(wrapped_add)  # Unknown | _Wrapped[..., Unknown, (*args: Unknown, **kwargs: Unknown), Unknown]
``` 

This can be fixed using `ParamSpec`:

```python
reveal_type(arapped_add)  # Type of "wrapped_add" is "(x: Tensor, y: Tensor) -> Tensor"
```

3. `GradScaler.scale` can accept either a `torch.Tensor` and return a `torch.Tensor` or accept an `Iterable[torch.Tensor]` and return an `Iterable[torch.Tensor]`. The signature does not differentiate when either is returned, and so implicitly returns the union of the two types (according to `pyright`). This leads to type errors when calling `.backward()` on a `Tensor` result:

```python
import torch
import torch.cuda.amp

grad_scaler = torch.cuda.amp.GradScaler()
loss = torch.tensor(1.0)

# Cannot access member "backward" for type "list[Unknown]"
# Cannot access member "backward" for type "tuple[Unknown, ...]"
# Cannot access member "backward" for type "map[Unknown]"
grad_scaler.scale(loss).backward()
```

This can be fixed by adding `@overload` definitions for `GradScaler.scale`:

```python
reveal_type(grad_scaler.scale(loss))  # Type of "grad_scaler.scale(loss)" is "Tensor"
reveal_type(grad_scaler.scale([loss, loss]))  # Type of "grad_scaler.scale([loss, loss])" is "List[Tensor]"
reveal_type(grad_scaler.scale((loss, loss)))  # Type of "grad_scaler.scale((loss, loss))" is "Tuple[Tensor, ...]"
reveal_type(grad_scaler.scale({loss, loss}))  # Type of "grad_scaler.scale({loss, loss})" is "Iterable[Tensor]"

```

### Versions

```console
$ python collect_env.py
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.20.5
Libc version: glibc-2.31

Python version: 3.8.17 (default, Jul  5 2023, 21:04:15)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-76-generic-x86_64-with-glibc2.17
Is CUDA available: N/A
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: 
GPU 0: NVIDIA RTX A6000
GPU 1: NVIDIA RTX A6000
GPU 2: NVIDIA RTX A6000
GPU 3: NVIDIA RTX A6000

Nvidia driver version: 525.105.17
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   43 bits physical, 48 bits virtual
CPU(s):                          64
On-line CPU(s) list:             0-63
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           49
Model name:                      AMD Ryzen Threadripper 3970X 32-Core Processor
Stepping:                        0
Frequency boost:                 enabled
CPU MHz:                         2200.000
CPU max MHz:                     3700.0000
CPU min MHz:                     2200.0000
BogoMIPS:                        7400.62
Virtualization:                  AMD-V
L1d cache:                       1 MiB
L1i cache:                       1 MiB
L2 cache:                        16 MiB
L3 cache:                        128 MiB
NUMA node0 CPU(s):               0-63
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.4.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] torch==2.1.0a0+gitb1ddd5a
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.1.0a0+gitb1ddd5a           dev_0    <develop>
```

### PR DESCRIPTION
**Pull request by ringohoffman**
Fixes #108629

1. Add the following to their modules' `__all__` so that pyright considers them to be publicly exported:
* [`torch.autocast`](https://pytorch.org/docs/stable/amp.html#torch.autocast)
* [`torch.cuda.amp.GradScaler`](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler)
* [`torch.cuda.amp.autocast`](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.autocast)
* [`torch.cuda.amp.custom_fwd`](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.custom_fwd)
* [`torch.cuda.amp.custom_bwd`](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.custom_bwd)
2. Add `overload`s for `torch.cuda.amp.GradScaler.scale` to differentiate when a `torch.Tensor` is returned vs. an `Iterable[torch.Tensor]` is returned based on the type of the `outputs` parameter.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel

### CODE REVIEW

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 1]
```diff
@@ -288,28 +302,25 @@ def unscale_(self, optimizer: torch.optim.Optimizer) -> None:
         if future_handles:
             torch.futures.wait_all(future_handles)
 
-    def step(
-        self, optimizer: torch.optim.Optimizer, *args, **kwargs
-    ) -> Optional[float]:
-        return super().step(optimizer, *args, **kwargs)
```
Comment (ringohoffman) [inline torch/distributed/fsdp/sharded_grad_scaler.py:294]: This is redundant and equivalent to not defining this at all.

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 2]
```diff
@@ -288,28 +302,25 @@ def unscale_(self, optimizer: torch.optim.Optimizer) -> None:
...
-                self._growth_tracker = 0
+            successful = self._growth_tracker + 1
+            if successful == self._growth_interval:
+                self._scale *= self._growth_factor
+                self._growth_tracker.fill_(0)
             else:
                 self._growth_tracker = successful
 
-    def update(self, new_scale: Optional[Union[float, FloatTensor]] = None) -> None:
+    def update(self, new_scale: Optional[Union[float, torch.Tensor]] = None) -> None:
```
Comment (ringohoffman) [inline torch/distributed/fsdp/sharded_grad_scaler.py:323]: `torch.cuda.FloatTensor` is not importable, so type checkers will not be able to do any validation against it. We may as well just use `torch.Tensor`.

File: torch/distributed/fsdp/sharded_grad_scaler.py [hunk 3]
```diff
@@ -100,9 +99,25 @@ def __init__(
...
+
+    @overload
+    def scale(self, outputs: Iterable[torch.Tensor]) -> Iterable[torch.Tensor]:
+        ...
+
     def scale(
-        self, outputs: Union[torch.Tensor, List[torch.Tensor]]
-    ) -> Union[torch.Tensor, List[torch.Tensor]]:
+        self, outputs: Union[torch.Tensor, Iterable[torch.Tensor]]
+    ) -> Union[torch.Tensor, Iterable[torch.Tensor]]:
```
Comment (ringohoffman) [inline torch/distributed/fsdp/sharded_grad_scaler.py:120]: The `@overload`s here provide more specific return types than what was here before.

Comment (ringohoffman) [reply torch/distributed/fsdp/sharded_grad_scaler.py:120]:
```python
import torch
import torch.cuda.amp

grad_scaler = torch.cuda.amp.GradScaler()
loss = torch.tensor(1.0)

# Cannot access member "backward" for type "list[Unknown]"
# Cannot access member "backward" for type "tuple[Unknown, ...]"
# Cannot access member "backward" for type "map[Unknown]"
grad_scaler.scale(loss).backward()
```

and now:

```python
reveal_type(grad_scaler.scale(loss))  # Type of "grad_scaler.scale(loss)" is "Tensor"
reveal_type(grad_scaler.scale([loss, loss]))  # Type of "grad_scaler.scale([loss, loss])" is "List[Tensor]"
reveal_type(grad_scaler.scale((loss, loss)))  # Type of "grad_scaler.scale((loss, loss))" is "Tuple[Tensor, ...]"
reveal_type(grad_scaler.scale({loss, loss}))  # Type of "grad_scaler.scale({loss, loss})" is "Iterable[Tensor]"

```

**Comment by ringohoffman** [comment]
@ezyang @albanD any thoughts on this?

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**